R/graded.R

```diff
@@ -564,25 +564,26 @@ grade_if_equal <- function(
   local_options_waldo_compare()
 
   compare_msg <- tryCatch(
-    waldo::compare(x, y, tolerance = tolerance),
+    {
+      time_limit <- knitr::opts_chunk$get("exercise.timelimit") %||%
+        gradethis_default_learnr_options$exercise.timelimit
+      time_limit <- time_limit * 0.8
+      # If `waldo::compare()` takes too long to evaluate,
+      # just skip to the fallback (`identical()`)
+      setTimeLimit(elapsed = time_limit, transient = TRUE)
+      waldo::compare(x, y, tolerance = tolerance)
+    },
     error = function(e) {
       # https://github.com/brodieG/diffobj/issues/152#issuecomment-788083359
       # waldo::compare() calls diffobj::ses() — these functions try hard to create
-      # a usable diff to describe the differences. These filters below cover
-      # cases where the diff description throws an error, but we know they only
-      # arise when a difference has occurred. Since we aren't (currently)
+      # a usable diff to describe the differences.
+      # We're engaging in some off-label usage of these functions,
+      # so they will sometimes error when we give them unusual inputs.
+      # If these functions throw an error, we fall back to `identical()`.
+      # Since we aren't (currently)
       # interested in reporting the differences between `x` and `y`, we mark
-      # these as known to be different
-      if (grepl("Exceeded buffer for finding fake snake", e$message, fixed = TRUE)) {
-        "different"
-      } else if (grepl("reached theoretically unreachable branch 2", e$message, fixed = TRUE)) {
-        "different"
-      } else {
-        warning(
-          "Error in grade_if_equal(): ", deparse(e$call), ": ", e$message, call. = FALSE
-        )
-        capture_graded(grade_grading_problem(error = e))
-      }
+      # these cases as simply "different" if they aren't identical.
```

Comment:
Very minor edits to the comment:

```r
      # waldo::compare() takes into account a lot of the things we'd have to
      # think about in comparing two objects, but its goal is to create a
      # readable diff. Since we're engaging in some off-label usage of these
      # functions, they will sometimes error or take longer than desired when we
      # give them unusual inputs. In these cases, we fall back to `identical()`.
      # Since we aren't (currently) interested in reporting the differences
      # between `x` and `y`, we mark them "different" if they aren't identical.
```

Reply:
Very well put!